src/libsync/pushnotifications.cpp

```diff
@@ -0,0 +1,189 @@
+#include "pushnotifications.h"
+#include "creds/abstractcredentials.h"
+#include "account.h"
+
+namespace {
+static constexpr int MAX_ALLOWED_FAILED_AUTHENTICATION_ATTEMPTS = 3;
+}
+
+namespace OCC {
+
+Q_LOGGING_CATEGORY(lcPushNotifications, "nextcloud.sync.pushnotifications", QtInfoMsg)
+
+PushNotifications::PushNotifications(Account *account, QObject *parent)
+    : QObject(parent)
+    , _account(account)
+{
+}
+
+PushNotifications::~PushNotifications()
+{
+    closeWebSocket();
+}
+
+void PushNotifications::setup()
+{
+    _isReady = false;
+    _failedAuthenticationAttemptsCount = 0;
+    reconnectToWebSocket();
+}
+
+void PushNotifications::reconnectToWebSocket()
+{
+    closeWebSocket();
+    openWebSocket();
+}
+
+void PushNotifications::closeWebSocket()
+{
+    if (_webSocket) {
+        qCInfo(lcPushNotifications) << "Close websocket";
+        _webSocket->close();
+    }
+}
+
+void PushNotifications::onWebSocketConnected()
+{
+    qCInfo(lcPushNotifications) << "Connected to websocket";
+
+    connect(_webSocket, &QWebSocket::textMessageReceived, this, &PushNotifications::onWebSocketTextMessageReceived, Qt::UniqueConnection);
+
+    authenticateOnWebSocket();
+}
+
+void PushNotifications::authenticateOnWebSocket()
+{
+    const auto credentials = _account->credentials();
+    const auto username = credentials->user();
+    const auto password = credentials->password();
+
+    // Authenticate
+    _webSocket->sendTextMessage(username);
+    _webSocket->sendTextMessage(password);
+}
+
+void PushNotifications::onWebSocketDisconnected()
+{
+    qCInfo(lcPushNotifications) << "Disconnected from websocket";
+}
+
+void PushNotifications::onWebSocketTextMessageReceived(const QString &message)
+{
+    qCInfo(lcPushNotifications) << "Received push notification:" << message;
+
+    if (message == "notify_file") {
+        handleNotifyFile();
+    } else if (message == "notify_activity" || message == "notify_notification") {
+        handleNotification();
+    } else if (message == "authenticated") {
+        handleAuthenticated();
+    } else if (message == "err: Invalid credentials") {
+        handleInvalidCredentials();
+    }
+}
+
+void PushNotifications::onWebSocketError(QAbstractSocket::SocketError error)
+{
+    qCWarning(lcPushNotifications) << "Websocket error" << error;
+
+    // This error gets thrown in testSetup_maxConnectionAttemptsReached_deletePushNotifications after
+    // the second connection attempt. I have no idea why this happens. Maybe the socket gets not closed correctly?
+    // I think it's fine to ignore this error.
+    if (error == QAbstractSocket::UnfinishedSocketOperationError) {
+        return;
```

Comment:
Probably worth either logging in here or moving the warning above after that if.